cudapoa/samples/sample_cudapoa.cpp

```diff
@@ -124,22 +125,148 @@ void process_batch(Batch* batch, bool msa, bool print)
     }
 }
 
-void generate_window_data(const std::string& input_file, const int number_of_windows, const int max_sequences_per_poa,
-                          std::vector<std::vector<std::string>>& windows, BatchSize& batch_size)
+size_t estimate_max_poas(const BatchSize& batch_size, const bool banded_alignment, const bool msa_flag)
 {
-    parse_window_data_file(windows, input_file, number_of_windows); // Generate windows.
-    assert(get_size(windows) > 0);
+    int32_t matrix_sequence_dimension = banded_alignment ? CUDAPOA_BANDED_MAX_MATRIX_SEQUENCE_DIMENSION : batch_size.max_matrix_sequence_dimension;
+    int32_t matrix_graph_dimension    = banded_alignment ? batch_size.max_matrix_graph_dimension_banded : batch_size.max_matrix_graph_dimension;
+    int32_t max_nodes_per_window      = banded_alignment ? batch_size.max_nodes_per_window_banded : batch_size.max_nodes_per_window;
 
-    int32_t max_read_length = 0;
-    for (auto& window : windows)
+    // Initialize CUDAPOA batch object for batched processing of POAs on the GPU.
+    size_t total = 0, free = 0;
+    cudaMemGetInfo(&free, &total);
+    size_t mem_per_batch         = 0.9 * free; // Using 90% of GPU available memory for CUDAPOA batch.
+    const int32_t mismatch_score = -6, gap_score = -8, match_score = 8;
+
+    int64_t sizeof_ScoreT = use32bitScore(batch_size, gap_score, mismatch_score, match_score) ? 4 : 2;
+    int64_t sizeof_SizeT  = use32bitSize(batch_size, banded_alignment) ? 4 : 2;
+
+    // Calculate memory requirements for POA arrays
+    int64_t device_size_per_poa = 0;
+    int64_t input_size_per_poa  = batch_size.max_sequences_per_poa * batch_size.max_sequence_size;
+    int64_t output_size_per_poa = batch_size.max_concensus_size;
+    device_size_per_poa += output_size_per_poa * sizeof(uint8_t);                                                                                // output_details_d_->consensus
+    device_size_per_poa += (!msa_flag) ? output_size_per_poa * sizeof(uint16_t) : 0;                                                             // output_details_d_->coverage
```

Comment:
The size calculation logic overlaps with logic in calculate_space_per_poa in cudapoa/src/allocate_block.hpp
Would it help to split the space calculation into individual functions per detail so that you can reuse those there and here?

Reply:
(if the code cannot be reused)

Reply:
that would definitely follow DRY rule, but I was debating whether to keep this component for the arrays or just estimate based on score matrix size. I possibly keep the estimation as is and rewrite it without repeating the code.

Reply:
changed to minimize repeated code.

Reply:
This looks good!